fc/virtualproduct.py

```diff
@@ -104,23 +105,30 @@ def scale_usgs_collection2(data):
     return data.apply(scale_and_clip_dataarray, keep_attrs=True,
                       scale_factor=0.275, add_offset=-2000, clip_range=(0, 10000))
 
-def scale_and_clip_dataarray(dataarray: xr.DataArray, *, scale_factor=1, add_offset=0, clip_range=None,
-                             new_nodata=-999, new_dtype='int16'):
+
+def scale_and_clip_dataarray(
+    dataarray: xr.DataArray,
+    scale_factor=1,
+    add_offset=0,
+    clip_range=None,
+    new_nodata=-999,
+    new_dtype="int16",
+):
     orig_attrs = dataarray.attrs
-    nodata = dataarray.attrs['nodata']
+    nodata = dataarray.attrs["nodata"]
 
     mask = dataarray.data == nodata
 
     dataarray = dataarray * scale_factor + add_offset
 
     if clip_range is not None:
-        clip_min, clip_max = clip_range
-        dataarray.clip(clip_min, clip_max)
+        dataarray = dataarray.clip(*clip_range)
```

Comment:
Two things.

1. The code in this function can be simplified by using `dataarray.clip(*clip_range, keep_attrs=True)`. `xarray` didn't have this option two years ago when this code was written.

2. This is definitely a bug, the intention of the code is to clearly supposed to clip the range. However, I'm not at all confident that it's the correct thing to do (see above comment about spectral signature). I don't know that any sort of validation was done when it was implemented. I think this was written by @dunkgray who was just taking a very naive approach to make USGS SR look roughly like GA SR.

None of this affects GA FC, so if you're confident that this is well enough tested and returns accurate results, it doesn't affect me at all.

Reply:
I don't want to change this any further... I can see that the data type and attribute handling may be unnecessary, but don't have the bravery to refactor it now.